File: src/engine/gameEngine.js
const SUITS = ["spades", "hearts", "diamonds", "clubs"];
const RANKS = ["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"];
const JOKER = "JOKER";
const RANK_VALUES = {
  "2": 2,
  "3": 3,
  "4": 4,
  "5": 5,
  "6": 6,
  "7": 7,
  "8": 8,
  "9": 9,
  "10": 10,
  J: 11,
  Q: 12,
  K: 13,
  A: 14,
};

export const ROUNDS = [
  {
    handSize: 7,
    requirements: [
      { type: "set", size: 3 },
      { type: "set", size: 3 },
    ],
  },
  {
    handSize: 8,
    requirements: [
      { type: "set", size: 3 },
      { type: "run", size: 4 },
    ],
  },
  {
    handSize: 9,
    requirements: [
      { type: "run", size: 4 },
      { type: "run", size: 4 },
    ],
  },
  {
    handSize: 10,
    requirements: [
      { type: "set", size: 4 },
      { type: "set", size: 5 },
    ],
  },
  {
    handSize: 11,
    requirements: [
      { type: "run", size: 7 },
      { type: "set", size: 3 },
    ],
  },
  {
    handSize: 12,
    requirements: [
      { type: "set", size: 3 },
      { type: "set", size: 3 },
      { type: "run", size: 4 },
    ],
  },
  {
    handSize: 12,
    requirements: [
      { type: "run", size: 4 },
      { type: "run", size: 4 },
      { type: "set", size: 3 },
    ],
  },
];

function joinWithAnd(parts) {
  if (parts.length <= 1) return parts[0] ?? "";
  if (parts.length === 2) return `${parts[0]} and ${parts[1]}`;
  return `${parts.slice(0, -1).join(", ")}, and ${parts[parts.length - 1]}`;
}

export function formatRequirements(requirements = []) {
  const groups = [];
  for (const req of requirements) {
    const key = `${req.type}:${req.size}`;
    const existing = groups.find((group) => group.key === key);
    if (existing) {
      existing.count += 1;
    } else {
      groups.push({ key, req, count: 1 });
    }
  }
  const parts = groups.map(({ req, count }) => {
    const isSet = req.type === "set";
    const label = isSet ? `${req.size}-of-a-kind` : `${req.size}-card straight flush`;
    if (count === 1) return label;
    const plural = isSet ? "s" : "es";
    return `${count} ${label}${plural}`;
  });
  return joinWithAnd(parts);
}

let nextId = 1;

export class Card {
  constructor(rank, suit) {
    this.rank = rank;
    this.suit = suit;
    this.cid = nextId++;
  }

  short() {
    return `${this.rank}${this.suit}`;
  }

  isWild() {
    return this.rank === "2" || this.rank === JOKER;
  }

  isJoker() {
    return this.rank === JOKER;
  }

  isRed() {
    return this.suit === "hearts" || this.suit === "diamonds";
  }
}

export class Meld {
  constructor({ type, rank = null, suit = null, cards }) {
    this.type = type;
    this.rank = rank;
    this.suit = suit;
    this.cards = cards;
  }

  canAdd(card) {
    if (this.type === "run") {
      return canFormRun([...this.cards, card], false);
    }
    return canFormSet([...this.cards, card], false);
  }

  add(card) {
    this.cards.push(card);
    if (this.type === "run") {
      this.cards = getSortedMeldCards(this);
    }
  }
}

export class Player {
  constructor(name) {
    this.name = name;
    this.hand = [];
    this.melds = [];
    this.stagedMelds = [];
    this.hasLaidDown = false;
    this.totalScore = 0;
    this.lastDiscardedRank = null;
    this.lastDiscardedId = null;
    this.aiNoProgressTurns = 0;
    this.strategyFlags = new Set();
  }
}

function numDecksForPlayers(players) {
  if (players <= 0) {
    throw new Error("Players must be >= 1");
  }
  if (players <= 2) return 1;
  if (players <= 4) return 2;
  if (players <= 6) return 3;
  if (players <= 8) return 4;
  if (players <= 10) return 5;
  throw new Error("Max 10 players supported");
}

export class Deck {
  constructor(players = 2, jokersPerDeck = 2) {
    const decks = numDecksForPlayers(players);
    this.cards = [];
    for (let d = 0; d < decks; d += 1) {
      for (const suit of SUITS) {
        for (const rank of RANKS) {
          this.cards.push(new Card(rank, suit));
        }
      }
      for (let j = 0; j < jokersPerDeck; j += 1) {
        this.cards.push(new Card(JOKER, "joker"));
      }
    }
  }

  shuffle() {
    for (let i = this.cards.length - 1; i > 0; i -= 1) {
      const j = Math.floor(Math.random() * (i + 1));
      [this.cards[i], this.cards[j]] = [this.cards[j], this.cards[i]];
    }
  }

  deal(count) {
    if (count < 0 || count > this.cards.length) {
      throw new Error("Invalid deal size");
    }
    const hand = this.cards.slice(0, count);
    this.cards = this.cards.slice(count);
    return hand;
  }
}

function combinations(items, size) {
  const results = [];
  const combo = [];
  function backtrack(start, remaining) {
    if (remaining === 0) {
      results.push([...combo]);
      return;
    }
    for (let i = start; i <= items.length - remaining; i += 1) {
      combo.push(items[i]);
      backtrack(i + 1, remaining - 1);
      combo.pop();
    }
  }
  backtrack(0, size);
  return results;
}

function canFormSet(cards, requireHalfNatural) {
  const naturals = cards.filter((card) => !card.isWild());
  if (requireHalfNatural && naturals.length < Math.ceil(cards.length / 2)) return false;
  if (naturals.length === 0) return !requireHalfNatural;
  const rank = naturals[0].rank;
  return naturals.every((card) => card.rank === rank);
}

function canFormRun(cards, requireHalfNatural) {
  const naturals = cards.filter((card) => !card.isWild());
  if (requireHalfNatural && naturals.length < Math.ceil(cards.length / 2)) return false;
  let suit = null;
  const naturalValues = [];
  let hasAce = false;
  for (const card of naturals) {
    if (suit && card.suit !== suit) return false;
    suit = card.suit;
    naturalValues.push(RANK_VALUES[card.rank]);
    if (card.rank === "A") hasAce = true;
  }
  const naturalSet = new Set(naturalValues);
  if (naturalSet.size !== naturalValues.length) return false;
  if (naturals.length === 0) return true;
  const size = cards.length;
  const canFitRun = (values, allowAceLow) => {
    const startMin = allowAceLow ? 1 : 2;
    for (let start = startMin; start <= 14 - size + 1; start += 1) {
      const needed = new Set();
      for (let offset = 0; offset < size; offset += 1) {
        needed.add(start + offset);
      }
      if (values.every((value) => needed.has(value))) {
        return true;
      }
    }
    return false;
  };

  if (canFitRun(naturalValues, false)) return true;
  if (!hasAce) return false;
  const lowValues = naturalValues.map((value) => (value === 14 ? 1 : value));
  const lowSet = new Set(lowValues);
  if (lowSet.size !== lowValues.length) return false;
  return canFitRun(lowValues, true);
}

function makeMeld(requirement, cards) {
  if (requirement.type === "set") {
    const natural = cards.find((card) => !card.isWild());
    return new Meld({
      type: "set",
      rank: natural ? natural.rank : null,
      cards: [...cards],
    });
  }
  const natural = cards.find((card) => !card.isWild());
  return new Meld({
    type: "run",
    suit: natural ? natural.suit : null,
    cards: [...cards],
  });
}

function findMeldOptions(hand, requirement, maxSize = hand.length) {
  const options = [];
  const startSize = requirement.size;
  const endSize = Math.max(startSize, Math.min(maxSize, hand.length));
  for (let size = endSize; size >= startSize; size -= 1) {
    for (const combo of combinations(hand, size)) {
      if (requirement.type === "set" && canFormSet(combo, true)) {
        options.push(makeMeld(requirement, combo));
      }
      if (requirement.type === "run" && canFormRun(combo, true)) {
        options.push(makeMeld(requirement, combo));
      }
    }
  }
  return options;
}

function findMeldsForRequirements(hand, requirements, { useAll = false } = {}) {
  const options = requirements.map((req) => findMeldOptions(hand, req));
  const used = new Set();

  function backtrack(reqIndex) {
    if (reqIndex === requirements.length) {
      if (useAll && used.size !== hand.length) return null;
      return [];
    }
    for (const meld of options[reqIndex]) {
      const meldIds = meld.cards.map((card) => card.cid);
      if (meldIds.some((cid) => used.has(cid))) continue;
      meldIds.forEach((cid) => used.add(cid));
      const rest = backtrack(reqIndex + 1);
      if (rest) return [meld, ...rest];
      meldIds.forEach((cid) => used.delete(cid));
    }
    return null;
  }

  return backtrack(0);
}

export function findTwoMelds(hand) {
  return findMeldsForRequirements(hand, ROUNDS[0].requirements);
}

export function canLayDownWithCards(hand, requirements) {
  return Boolean(findMeldsForRequirements(hand, requirements));
}

export function canLayDownWithCard(hand, card, requirements) {
  return Boolean(findMeldsForRequirements([...hand, card], requirements));
}

export class Game {
  constructor(players = 2, dealerIndex = 0) {
    if (players < 2 || players > 10) {
      throw new Error("Players must be between 2 and 10.");
    }
    this.players = Array.from({ length: players }, (_, idx) => new Player(`Player ${idx + 1}`));
    this.dealerIndex = dealerIndex;
    this.currentPlayerIndex = 0;
    this.drawPile = [];
    this.discardPile = [];
    this.deadPile = [];
    this.playersCount = players;
    this.roundIndex = 0;
    this.startRound();
  }

  startRound() {
    const deck = new Deck(this.playersCount);
    deck.shuffle();

    for (const player of this.players) {
      player.hand = [];
      player.melds = [];
      player.stagedMelds = [];
      player.hasLaidDown = false;
      player.aiNoProgressTurns = 0;
    }
    this.deadPile = [];

    const order = [...Array(this.players.length).keys()];
    const start = (this.dealerIndex + 1) % this.players.length;
    const dealOrder = order.slice(start).concat(order.slice(0, start));

    const currentRound = this.currentRound();
    for (let round = 0; round < currentRound.handSize; round += 1) {
      for (const idx of dealOrder) {
        this.players[idx].hand.push(...deck.deal(1));
      }
    }

    this.drawPile = [...deck.cards];
    this.discardPile = [];
    if (this.drawPile.length > 0) {
      this.discardPile.push(this.drawPile.shift());
    }
    this.currentPlayerIndex = start;
  }

  currentRound() {
    return ROUNDS[this.roundIndex] ?? ROUNDS[0];
  }

  nextRound() {
    if (this.roundIndex < ROUNDS.length - 1) {
      this.roundIndex += 1;
    }
    this.startRound();
  }

  currentPlayer() {
    return this.players[this.currentPlayerIndex];
  }

  otherPlayer() {
    return this.players[(this.currentPlayerIndex + 1) % this.players.length];
  }

  drawFromDiscard(player) {
    if (this.discardPile.length === 0) return null;
    const card = this.discardPile.pop();
    player.hand.push(card);
    return card;
  }

  drawFromStock(player) {
    if (this.drawPile.length === 0) {
      this.reshuffleNonHandCards();
    }
    if (this.drawPile.length === 0) return null;
    const card = this.drawPile.shift();
    player.hand.push(card);
    return card;
  }

  discard(player, card) {
    player.hand = player.hand.filter((c) => c.cid !== card.cid);
    player.lastDiscardedRank = card.rank;
    player.lastDiscardedId = card.cid;
    if (this.discardPile.length > 0) {
      this.deadPile.push(...this.discardPile);
      this.discardPile = [];
    }
    this.discardPile.push(card);
  }

  tryLayDown(player) {
    if (player.hasLaidDown) return false;
    const melds = findMeldsForRequirements(player.hand, this.currentRound().requirements);
    if (!melds) return false;
    for (const meld of melds) {
      for (const card of meld.cards) {
        player.hand = player.hand.filter((c) => c.cid !== card.cid);
      }
      player.melds.push(meld);
    }
    player.hasLaidDown = true;
    return true;
  }

  tryLayDownWithCards(player, cards) {
    if (player.hasLaidDown) return false;
    const requirements = this.currentRound().requirements;
    const handIds = new Set(player.hand.map((card) => card.cid));
    if (!cards.every((card) => handIds.has(card.cid))) return false;
    const melds = findMeldsForRequirements(cards, requirements, { useAll: true });
    if (!melds) return false;
    for (const meld of melds) {
      for (const card of meld.cards) {
        player.hand = player.hand.filter((c) => c.cid !== card.cid);
      }
      player.melds.push(meld);
    }
    player.hasLaidDown = true;
    return true;
  }

  stageCard(player, card, meldIndex = null) {
    const handHasCard = player.hand.some((c) => c.cid === card.cid);
    if (!handHasCard) return false;
    player.hand = player.hand.filter((c) => c.cid !== card.cid);
    if (meldIndex !== null && player.stagedMelds[meldIndex]) {
      player.stagedMelds[meldIndex].cards.push(card);
      return true;
    }
    player.stagedMelds.push({ cards: [card], staged: true });
    return true;
  }

  unstageCard(player, card) {
    for (let i = 0; i < player.stagedMelds.length; i += 1) {
      const meld = player.stagedMelds[i];
      const idx = meld.cards.findIndex((c) => c.cid === card.cid);
      if (idx === -1) continue;
      meld.cards.splice(idx, 1);
      if (meld.cards.length === 0) {
        player.stagedMelds.splice(i, 1);
      }
      player.hand.push(card);
      return true;
    }
    return false;
  }

  clearStaged(player) {
    for (const meld of player.stagedMelds) {
      player.hand.push(...meld.cards);
    }
    player.stagedMelds = [];
  }

  tryLayDownStaged(player) {
    if (player.hasLaidDown) return false;
    const stagedMelds = player.stagedMelds;
    const requirements = this.currentRound().requirements;
    if (stagedMelds.length !== requirements.length) {
      this.clearStaged(player);
      return false;
    }

    const used = new Set();
    const assignment = [];

    function matchesRequirement(meld, req) {
      if (meld.cards.length < req.size) return false;
      if (req.type === "set") return canFormSet(meld.cards, true);
      return canFormRun(meld.cards, true);
    }

    function backtrack(reqIndex) {
      if (reqIndex === requirements.length) return true;
      for (let i = 0; i < stagedMelds.length; i += 1) {
        if (used.has(i)) continue;
        const requirement = requirements[reqIndex];
        if (!matchesRequirement(stagedMelds[i], requirement)) continue;
        used.add(i);
        assignment.push({ meld: stagedMelds[i], requirement });
        if (backtrack(reqIndex + 1)) return true;
        assignment.pop();
        used.delete(i);
      }
      return false;
    }

    if (!backtrack(0)) {
      this.clearStaged(player);
      return false;
    }

    for (const { meld, requirement } of assignment) {
      const committed = makeMeld({ type: requirement.type }, meld.cards);
      player.melds.push(committed);
    }
    player.stagedMelds = [];
    player.hasLaidDown = true;
    return true;
  }

  autoStageMelds(player) {
    if (player.hasLaidDown) return false;
    this.clearStaged(player);
    const melds = findMeldsForRequirements(player.hand, this.currentRound().requirements);
    if (!melds) return false;
    for (const meld of melds) {
      for (const card of meld.cards) {
        player.hand = player.hand.filter((c) => c.cid !== card.cid);
      }
      player.stagedMelds.push({
        type: meld.type,
        rank: meld.rank ?? null,
        suit: meld.suit ?? null,
        cards: [...meld.cards],
        staged: true,
      });
    }
    return true;
  }

  layOffAll(player) {
    if (!player.hasLaidDown) return 0;
    const allMelds = this.players.flatMap((p) => p.melds);
    if (allMelds.length === 0) return 0;
    let moved = 0;
    const handSorted = [...player.hand].sort((a, b) => Number(a.isWild()) - Number(b.isWild()));
    for (const card of handSorted) {
      if (player.hand.length <= 1) break;
      for (const meld of allMelds) {
        if (meld.canAdd(card)) {
          meld.add(card);
          player.hand = player.hand.filter((c) => c.cid !== card.cid);
          moved += 1;
          break;
        }
      }
    }
    return moved;
  }

  layOffCardToMeld(player, card, meld) {
    const playerHasCard = player.hand.some((c) => c.cid === card.cid);
    const melds = this.players.flatMap((p) => p.melds);
    if (!playerHasCard) return false;
    if (!player.hasLaidDown) return false;
    if (!melds.includes(meld)) return false;
    if (!meld.canAdd(card)) return false;
    meld.add(card);
    player.hand = player.hand.filter((c) => c.cid !== card.cid);
    return true;
  }

  canLayOff(player) {
    if (!player.hasLaidDown) return false;
    const allMelds = this.players.flatMap((p) => p.melds);
    if (allMelds.length === 0) return false;
    if (player.hand.length <= 1) return false;
    return player.hand.some((card) => allMelds.some((meld) => meld.canAdd(card)));
  }

  canLayOffCard(card) {
    return this.players.some((player) => player.melds.some((meld) => meld.canAdd(card)));
  }

  meldRanksFor(playerIndex) {
    const ranks = new Set();
    for (const meld of this.players[playerIndex].melds) {
      if (meld.type === "set" && meld.rank) {
        ranks.add(meld.rank);
        continue;
      }
      for (const card of meld.cards) {
        if (!card.isWild()) {
          ranks.add(card.rank);
        }
      }
    }
    return ranks;
  }

  opponentMeldRanks(playerIndex) {
    const ranks = new Set();
    for (let idx = 0; idx < this.players.length; idx += 1) {
      if (idx === playerIndex) continue;
      for (const meld of this.players[idx].melds) {
        if (meld.type === "set" && meld.rank) {
          ranks.add(meld.rank);
          continue;
        }
        for (const card of meld.cards) {
          if (!card.isWild()) {
            ranks.add(card.rank);
          }
        }
      }
    }
    return ranks;
  }

  handPoints(hand) {
    let total = 0;
    for (const card of hand) {
      if (card.isWild()) {
        total += 20;
      } else if (["10", "J", "Q", "K", "A"].includes(card.rank)) {
        total += 10;
      } else {
        total += 5;
      }
    }
    return total;
  }

  applyRoundScores(winnerIndex) {
    const roundScores = [];
    for (let idx = 0; idx < this.players.length; idx += 1) {
      const player = this.players[idx];
      const roundScore = idx === winnerIndex ? 0 : this.handPoints(player.hand);
      player.totalScore += roundScore;
      roundScores.push(roundScore);
    }
    return roundScores;
  }

  checkWinAfterDiscard(player) {
    return player.hand.length === 0;
  }

  checkWin(player) {
    return player.hand.length === 0;
  }

  reshuffleNonHandCards() {
    const pool = [];
    const topDiscard = this.discardPile.length > 0 ? this.discardPile[this.discardPile.length - 1] : null;
    if (this.discardPile.length > 1) {
      pool.push(...this.discardPile.slice(0, -1));
    }
    if (this.deadPile.length > 0) {
      pool.push(...this.deadPile);
    }
    this.deadPile = [];
    this.discardPile = topDiscard ? [topDiscard] : [];
    for (let i = pool.length - 1; i > 0; i -= 1) {
      const j = Math.floor(Math.random() * (i + 1));
      [pool[i], pool[j]] = [pool[j], pool[i]];
    }
    this.drawPile = pool;
    if (this.discardPile.length === 0 && this.drawPile.length > 0) {
      this.discardPile.push(this.drawPile.shift());
    }
  }
}

export const SuitSymbols = {
  spades: "&spades;",
  hearts: "&hearts;",
  diamonds: "&diams;",
  clubs: "&clubs;",
  joker: "J",
};

export const Ranks = RANKS;
export const JokerRank = JOKER;

/**
 * Returns meld cards sorted by effective rank position.
 * For sets: naturals first, then wilds.
 * For runs: cards sorted by their effective rank in the sequence,
 *           with wilds assigned to fill gaps.
 */
export function getSortedMeldCards(meld) {
  if (!meld || !Array.isArray(meld.cards)) return [];
  if (!meld.type) {
    return [...meld.cards];
  }
  const isWild = (card) => (card?.isWild ? card.isWild() : card?.rank === "2" || card?.rank === JOKER);
  if (meld.type === "set") {
    // Sets: naturals first, then wilds
    const naturals = meld.cards.filter((c) => !isWild(c));
    const wilds = meld.cards.filter((c) => isWild(c));
    return [...naturals, ...wilds];
  }

  // Run: sort by effective position in the straight
  const cards = [...meld.cards];
  const naturals = cards.filter((c) => !isWild(c));
  const wilds = cards.filter((c) => isWild(c));

  if (naturals.length === 0) {
    // All wilds - just return as-is
    return cards;
  }

  // Find the natural card values
  const naturalValues = naturals.map((c) => RANK_VALUES[c.rank]);
  const minNatural = Math.min(...naturalValues);
  const maxNatural = Math.max(...naturalValues);

  // Determine the run range - try to find smallest valid range
  const size = cards.length;
  let runStart = null;
  for (let start = 2; start <= 14 - size + 1; start += 1) {
    const rangeEnd = start + size - 1;
    if (minNatural >= start && maxNatural <= rangeEnd) {
      runStart = start;
      break;
    }
  }

  if (runStart === null) {
    // Shouldn't happen for valid meld, return as-is
    return cards;
  }

  // Build sorted array: assign each position in the run
  const sorted = [];
  const usedNaturals = new Set();
  const wildQueue = [...wilds];

  for (let pos = runStart; pos < runStart + size; pos += 1) {
    const natural = naturals.find(
      (c) => RANK_VALUES[c.rank] === pos && !usedNaturals.has(c.cid)
    );
    if (natural) {
      sorted.push(natural);
      usedNaturals.add(natural.cid);
    } else if (wildQueue.length > 0) {
      sorted.push(wildQueue.shift());
    }
  }

  return sorted;
}

File: game_engine.py
"""Core game logic for Chinese Rummy mini-game 1."""

from __future__ import annotations

import random
from dataclasses import dataclass, field
from itertools import combinations, count

SUITS = ["♠", "♥", "♦", "♣"]
RANKS = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"]
JOKER = "🃏"


def num_decks_for_players(players: int) -> int:
    if players <= 0:
        raise ValueError("Players must be >= 1")
    if players <= 2:
        return 1
    if players <= 4:
        return 2
    if players <= 6:
        return 3
    if players <= 8:
        return 4
    if players <= 10:
        return 5
    raise ValueError("Max 10 players supported")


_id_counter = count(1)


@dataclass(frozen=True)
class Card:
    rank: str
    suit: str
    cid: int = field(default_factory=lambda: next(_id_counter))

    def short(self) -> str:
        return f"{self.rank}{self.suit}"

    def is_wild(self) -> bool:
        return self.rank == "2" or self.rank == JOKER

    def is_joker(self) -> bool:
        return self.rank == JOKER

    def is_red(self) -> bool:
        return self.suit in {"♥", "♦"}


class Deck:
    def __init__(self, players: int = 2, jokers_per_deck: int = 2) -> None:
        decks = num_decks_for_players(players)
        cards: list[Card] = []
        for _ in range(decks):
            cards.extend(Card(rank, suit) for suit in SUITS for rank in RANKS)
            cards.extend(Card(JOKER, JOKER) for _ in range(jokers_per_deck))
        self.cards = cards

    def shuffle(self) -> None:
        random.shuffle(self.cards)

    def deal(self, n: int) -> list[Card]:
        if n < 0 or n > len(self.cards):
            raise ValueError("Invalid deal size")
        hand, self.cards = self.cards[:n], self.cards[n:]
        return hand


@dataclass
class Meld:
    rank: str
    cards: list[Card]

    def can_add(self, card: Card) -> bool:
        return card.is_wild() or card.rank == self.rank

    def add(self, card: Card) -> None:
        self.cards.append(card)


@dataclass
class Player:
    name: str
    hand: list[Card] = field(default_factory=list)
    melds: list[Meld] = field(default_factory=list)
    has_laid_down: bool = False
    total_score: int = 0


def _meld_candidates(hand: list[Card]) -> list[tuple[list[Card], str]]:
    candidates: list[tuple[list[Card], str]] = []
    for trio in combinations(hand, 3):
        wilds = [c for c in trio if c.is_wild()]
        naturals = [c for c in trio if not c.is_wild()]
        if len(naturals) < 2:
            continue
        ranks = {c.rank for c in naturals}
        if len(ranks) != 1:
            continue
        if len(wilds) > 1:
            continue
        rank = next(iter(ranks))
        candidates.append((list(trio), rank))
    return candidates


def find_two_melds(hand: list[Card]) -> list[Meld] | None:
    candidates = _meld_candidates(hand)
    for first_cards, first_rank in candidates:
        first_ids = {c.cid for c in first_cards}
        for second_cards, second_rank in candidates:
            second_ids = {c.cid for c in second_cards}
            if first_ids.isdisjoint(second_ids):
                return [
                    Meld(rank=first_rank, cards=first_cards),
                    Meld(rank=second_rank, cards=second_cards),
                ]
    return None


class Game:
    def __init__(self, players: int = 2, dealer_index: int = 0) -> None:
        if players != 2:
            raise ValueError("Prototype supports 2 players only.")
        self.players: list[Player] = [Player(name="You"), Player(name="Opponent")]
        self.dealer_index = dealer_index
        self.current_player_index = 0
        self.draw_pile: list[Card] = []
        self.discard_pile: list[Card] = []
        self.players_count = players
        self.start_round()

    def start_round(self) -> None:
        deck = Deck(players=self.players_count)
        deck.shuffle()

        for player in self.players:
            player.hand.clear()
            player.melds.clear()
            player.has_laid_down = False

        order = list(range(len(self.players)))
        start = (self.dealer_index + 1) % len(self.players)
        order = order[start:] + order[:start]

        for _ in range(7):
            for idx in order:
                self.players[idx].hand.extend(deck.deal(1))

        self.draw_pile = deck.cards[:]
        self.discard_pile = []
        if self.draw_pile:
            self.discard_pile.append(self.draw_pile.pop(0))
        self.current_player_index = start

    def hand_points(self, hand: list[Card]) -> int:
        total = 0
        for card in hand:
            if card.is_wild():
                total += 20
            elif card.rank in {"10", "J", "Q", "K", "A"}:
                total += 10
            else:
                total += 5
        return total

    def apply_round_scores(self, winner_index: int) -> list[int]:
        round_scores: list[int] = []
        for idx, player in enumerate(self.players):
            if idx == winner_index:
                round_score = 0
            else:
                round_score = self.hand_points(player.hand)
            player.total_score += round_score
            round_scores.append(round_score)
        return round_scores

    def current_player(self) -> Player:
        return self.players[self.current_player_index]

    def other_player(self) -> Player:
        return self.players[(self.current_player_index + 1) % len(self.players)]

    def draw_from_discard(self, player: Player) -> Card | None:
        if not self.discard_pile:
            return None
        card = self.discard_pile.pop()
        player.hand.append(card)
        return card

    def draw_from_stock(self, player: Player) -> Card | None:
        if not self.draw_pile:
            self._reshuffle_non_hand_cards()
        if not self.draw_pile:
            return None
        card = self.draw_pile.pop(0)
        player.hand.append(card)
        return card

    def discard(self, player: Player, card: Card) -> None:
        player.hand = [c for c in player.hand if c.cid != card.cid]
        self.discard_pile.append(card)

    def try_lay_down(self, player: Player) -> bool:
        if player.has_laid_down:
            return False
        melds = find_two_melds(player.hand)
        if not melds:
            return False
        for meld in melds:
            for card in meld.cards:
                player.hand = [c for c in player.hand if c.cid != card.cid]
            player.melds.append(meld)
        player.has_laid_down = True
        return True

    def lay_off_all(self, player: Player) -> int:
        if not player.has_laid_down and not player.melds:
            return 0
        all_melds: list[Meld] = []
        for p in self.players:
            all_melds.extend(p.melds)
        if not all_melds:
            return 0
        moved = 0
        hand_sorted = sorted(player.hand, key=lambda c: c.is_wild())
        for card in list(hand_sorted):
            if len(player.hand) <= 1:
                break
            for meld in all_melds:
                if meld.can_add(card):
                    meld.add(card)
                    player.hand = [c for c in player.hand if c.cid != card.cid]
                    moved += 1
                    break
        return moved

    def lay_off_card_to_meld(self, player: Player, card: Card, meld: Meld) -> bool:
        if card not in player.hand:
            return False
        if meld not in [m for p in self.players for m in p.melds]:
            return False
        if not meld.can_add(card):
            return False
        meld.add(card)
        player.hand = [c for c in player.hand if c.cid != card.cid]
        return True

    def can_lay_off(self, player: Player) -> bool:
        all_melds: list[Meld] = []
        for p in self.players:
            all_melds.extend(p.melds)
        if not all_melds:
            return False
        if len(player.hand) <= 1:
            return False
        for card in player.hand:
            for meld in all_melds:
                if meld.can_add(card):
                    return True
        return False

    def can_lay_off_card(self, card: Card) -> bool:
        for p in self.players:
            for meld in p.melds:
                if meld.can_add(card):
                    return True
        return False

    def meld_ranks_for(self, player_index: int) -> set[str]:
        ranks: set[str] = set()
        for meld in self.players[player_index].melds:
            ranks.add(meld.rank)
        return ranks

    def opponent_meld_ranks(self, player_index: int) -> set[str]:
        ranks: set[str] = set()
        for idx, player in enumerate(self.players):
            if idx == player_index:
                continue
            for meld in player.melds:
                ranks.add(meld.rank)
        return ranks

    def check_win_after_discard(self, player: Player) -> bool:
        return len(player.hand) == 0

    def _reshuffle_non_hand_cards(self) -> None:
        pool: list[Card] = []
        for player in self.players:
            for meld in player.melds:
                pool.extend(meld.cards)
            player.melds.clear()
            player.has_laid_down = False
        pool.extend(self.discard_pile)
        self.discard_pile = []
        random.shuffle(pool)
        self.draw_pile = pool
        if self.draw_pile:
            self.discard_pile.append(self.draw_pile.pop(0))


def ai_choose_discard(
    hand: list[Card],
    can_lay_down: bool,
    can_lay_off: bool,
    avoid_ranks: set[str],
    keep_ranks: set[str],
) -> Card:
    # Point-aware but conservative: avoid discarding wilds unless forced.
    non_wild = [c for c in hand if not c.is_wild()]
    if non_wild:
        counts: dict[str, int] = {}
        for card in non_wild:
            counts[card.rank] = counts.get(card.rank, 0) + 1
        # Protect any card that would complete a pair to a trip.
        protected = {c.rank for c in non_wild if counts[c.rank] >= 2}
        candidates = [
            c
            for c in non_wild
            if c.rank not in protected and c.rank not in avoid_ranks and c.rank not in keep_ranks
        ]
        if not candidates:
            candidates = [c for c in non_wild if c.rank not in protected and c.rank not in keep_ranks]
        if not candidates:
            candidates = non_wild

        # Prefer singletons first, then lower counts; within that, drop higher point cards.
        def card_points(c: Card) -> int:
            if c.rank in {"10", "J", "Q", "K", "A"}:
                return 10
            return 5

        candidates.sort(key=lambda c: (counts[c.rank], -card_points(c), c.rank))
        return candidates[0]
    # If all wild (rare), discard the first.
    return hand[0]
